### PR TITLE
Allow partial deployments when sources are missing locally

### PR DESCRIPTION
**Pull request by Arkham**
### WHY are these changes introduced?

Fixes https://github.com/Shopify/internal-cli-foundations/issues/384

### WHAT is this pull request doing?

- Always ask for confirmation before deploying (user can provide `--force` to skip the confirmation prompt)
- Relax constraints so that the deploy can proceed even if an extension (or function) is missing locally
- Use new ink-powered prompts to display a summary for the user

### How to test your changes?

- create a new app: `pnpm create-app --path ~/Desktop`
- cd into app and create an extension `pnpm generate extension`
- open a new terminal side by side, go into the CLI folder and run `pnpm shopify app deploy --path ~/Desktop/new-app`
- you should see the confirmation prompt that you want to create a new extension
- type `n` to cancel
- run `pnpm shopify app deploy --path ~/Desktop/new-app --force` to deploy without any prompts
- in the app terminal, create a new extension `pnpm generate extension`
- in the CLI terminal,  and run `pnpm shopify app deploy --path ~/Desktop/new-app`
- you should see the prompt that you want to **create** a new extension and **update** an existing extension
- type 'y' to proceed
- in the app terminal, delete one of the extensions
- in the CLI terminal,  run `pnpm shopify app deploy --path ~/Desktop/new-app`
- you should see the prompt that you want to **update** an extension and that another extension is **missing locally**

### Post-release steps

- Coordinate with @shainaraskas for documentation changes

### Measuring impact

How do we know this change was effective? Please choose one:

- [x] n/a - this doesn't need measurement, e.g. a linting rule or a bug-fix
- [ ] Existing analytics will cater for this addition
- [ ] PR includes analytics changes to measure impact

### Checklist

- [x] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [x] I've considered possible [documentation](https://shopify.dev) changes
- [x] I've made sure that any changes to `dev` or `deploy` have been reflected in the [internal flowchart](https://www.figma.com/file/7vqUp50u6dm48Zfb4JRRn8/CLI3-Internals).

### CODE REVIEW

File: packages/app/src/cli/services/environment/identifiers.ts [hunk 1]
```diff
@@ -12,6 +12,7 @@ export interface EnsureDeploymentIdsPresenceOptions {
   appId: string
   appName: string
   envIdentifiers: Partial<Identifiers>
+  force: boolean
```
Comment (isaacroldan) [inline packages/app/src/cli/services/environment/identifiers.ts:15]: In this same file, should we remove the `environment.local.useFunctionMatching()` check and enable function matching always? Don't think it makes sense to keep it under a flag now that the deploy is not strict

Comment (Arkham) [reply packages/app/src/cli/services/environment/identifiers.ts:15]: sure thing, thanks!

**Review by isaacroldan — APPROVED** [review]

**Review by matteodepalo — APPROVED** [review]

**Comment by matteodepalo** [comment]
@Arkham did you coordinate with @shainaraskas for doc changes? I'm going to remove the post-release-steps label.

**Comment by Arkham** [comment]
I forgot, thanks for reminding me! Just opened this one: https://github.com/Shopify/shopify-dev/pull/30167

**Pull request merged**